### DISCUSSION TITLE
SGP41 (SGP4X) Sensors Support (VOC and NOx)

### DISCUSSION THREAD

**Discussion by precurse** [arendst/Tasmota#18867] [Ideas]
I added support for the SGP41 VOC/NOx sensor to Tasmota (https://github.com/arendst/Tasmota/compare/development...precurse:Tasmota:sgp41) 

It uses the official Sensirion driver, which is more up-to-date compared with the previous Adafruit SGP40 driver. It works well, but requires recompilation to support it at this time. The issue is that both the SGP40 and SGP41 use the same I2C device identifier, and I don't know of a way to detect it. I've tried to first detect the SGP40, by connecting to it (see below), then run an I2C command. If that command fails, then I want it to fallback to the SGP41. However, if I try it this way, the whole I2C bus craps out and my GPIO configuration resets to "None". I assume it has to do with the 2 calls to `sgp40.Begin(Wire)` and then `sgp41.Begin(Wire)` a few lines after that.

```
  // First try SGP40
   sgp40.begin(Wire);
   error = sgp40.getSerialNumber(serialNumber, serialNumberSize);

   if (!error) {
     // hacky way to detect SGP40
     error = sgp40.measureRawSignal(0x8000, 0x6666, srawVoc);

     if (!error) {
       // We got an SGP40
       sgp4x_found = true;
       sgp4x_type = Sgp4x_types::SGP40; 
     } else {
       // Continue to test for SGP41
     }
   }

 if (!sgp4x_found) {
    // Try to detect SGP41
    sgp41.begin(Wire);
    error = sgp41.getSerialNumber(serialNumber, serialNumberSize);

    if (error) {
      return;
    } else {
      // hacky way to detect SGP41
      error = sgp41.executeConditioning(0x8000, 0x6666, srawVoc);

      if (!error) {
        // We found an SGP41
        sgp4x_found = true;
        sgp4x_type = Sgp4x_types::SGP41; 
      } else {
        // No SGP41 or SGP40 detected
        return;
      }
    }
  }
... omitted for brevity...
```

If anyone has an idea of how I can handle this, it would be appreciated. Thanks!

**Comment by barbudor** [comment]
Why not just creating a different driver with a different index, a different USE_xxx
Up to users to enable one or the other with I2cdriver command for just any other conflict when the same address is used by different sensors

**Reply by precurse** [reply]
That's definitely an option. The SGP40 and SGP41 are so similar though, and was hoping to have them as the same `i2cdriver` to make it easier for auto detection.

Thanks for the pointer though. If I can't find a workaround I may just do that.

**Comment by TMHEINE** [comment]
I found an mistake in code for sensor "xsns_109_sgp4x.ino"
in MQTT SGP40 is transmitted instead of SGP41.
I think the mistake is here:

`void`` Sgp4xShow(bool json)
{
  if (sgp4x_state == STATE_SGP4X_NORMAL) {
    if (json) {
      if (sgp4x_type == TYPE_SGP41) {
        ResponseAppend_P(PSTR(",\"SGP40\":{\"VOC_" D_JSON_RAW "\":%d,\"NOX_" D_JSON_RAW "\":%d,\"" D_TVOC "\":%d,\"" D_NOX "\":%d"), srawVoc, srawNox, voc_index_sgp4x, nox_index_sgp4x);
      } else {
        ResponseAppend_P(PSTR(",\"SGP41\":{\"VOC_" D_JSON_RAW "\":%d,,\"" D_TVOC "\":%d,"), srawVoc, voc_index_sgp4x);
      }
      ResponseJsonEnd();
#ifdef USE_DOMOTICZ
      if (0 == TasmotaGlobal.tele_period) DomoticzSensor(DZ_AIRQUALITY, srawVoc);
`

Means in line 268 and 270. 
The names of de selected sensors are crossed.

So it shows SGP41 when having an SGP40 and the other way round.

**Reply by barbudor** [reply]
Thanks
Can you make a PR ?

**Reply by precurse** [reply]
Thanks. I created a PR for this here https://github.com/arendst/Tasmota/pull/20439
